### PR TITLE
[MRG] Fix type column checking in *_channels.tsv files

### LINKED ISSUE #1018: Validate "type" column in *_channels.tsv
Hello, while working with the  [OpenNeuro ds001810 dataset](https://openneuro.org/datasets/ds001810) (EEG data), I discovered that e.g. `sub-01/ses-anodalpre/eeg/sub-01_ses-anodalpre_task-attentionalblink_channels.tsv` looks like this:

```
name    type    units   sampling_frequency      description     status  status_description
...
PO4     eeg     uV      512     EEG channel     good    n/a
O2      eeg     uV      512     EEG channel     good    n/a
EXG1    eog     uV      512     above left eyebrow (VEOG1)      good    n/a
EXG2    eog     uV      512     below left eye (VEOG2)  good    n/a
EXG3    misc    uV      512     left earlobe (for re-referencing)       good    n/a
EXG4    misc    uV      512     right earlobe (for re-referencing)      good    n/a
...
Status  trig    unknown 512     triggers        good    n/a
```
(`...` added by me to indicate I have omitted lines)
As you can see, the values in the `type` column are all in lowercase letters. **This dataset passes validation.**

However, the [BIDS specification for EEG data](https://bids-specification.readthedocs.io/en/latest/04-modality-specific-files/03-electroencephalography.html) demands **uppercase** spelling. So I suppose this is a rule that is currently not checked or enforced by the validator.

This has already caused an [issue](https://github.com/mne-tools/mne-bids/issues/481) for us downstream. I would therefore appreciate if this could be addressed.

Thanks for your help,

–Richard.

### PR DESCRIPTION
**Pull request by adam2392**
Closes: #1018 

Adds:

- `checkTypeCol.js` for M/IEEG data
-  unit tests added 
-  unit tests updated to reflect new Correct state for `type` column

### CODE REVIEW

File: bids-validator/utils/issues/list.js [hunk 1]
```diff
@@ -713,5 +713,15 @@ export default {
     key: 'SCANS_FILENAME_NOT_MATCH_DATASET',
     severity: 'error',
     reason: 'The filename in scans.tsv file does not match what is present in the BIDS dataset.'
+  },
+  130: {
+    key: 'CHANNELS_COLUMN_TYPE_UPPER_CASE',
+    severity: 'warning',
```
Comment (sappelhoff) [inline bids-validator/utils/issues/list.js]: okay with me to keep this a warning for now --> but if we decide in https://github.com/bids-standard/bids-specification/issues/534 that it MUST be upper case, we can post-hoc change the "severity" to error.

File: bids-validator/utils/issues/list.js [hunk 2]
```diff
@@ -713,5 +713,15 @@ export default {
...
+    reason: 'Type column in channels.tsv files should consist of upper-case characters.'
+  },
+  131: {
+    key: 'CHANNELS_COLUMN_TYPE',
+    severity: 'error',
+    reason: 'Type column in channels.tsv files should only consist of values allowed in the specification for (M)(i)EEG data.'
```
Comment (sappelhoff) [inline bids-validator/utils/issues/list.js]:
```suggestion
    reason: 'Type column in channels.tsv files should only consist of values allowed in the specification for MEG/EEG/iEEG data.'
```

File: bids-validator/validators/tsv/checkStatusCol.js [hunk 3]
```diff
@@ -34,4 +34,5 @@ const checkStatusCol = function(rows, file, issues) {
   return
 }
 
-export default checkStatusCol
+
+export default checkStatusCol
```
Comment (sappelhoff) [inline bids-validator/validators/tsv/checkStatusCol.js]: why this change?

File: bids-validator/validators/tsv/checkTypeCol.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
+const Issue = require('../../utils').issues.Issue
+
+// allowable 'type' values from the BIDS specification
+const allowedMEEGTypes = ['EEG', 'ECOG', 'SEEG', 'DBS', 'PD', 'OTHER',  // (i)EEG
+    'EOG', 'ECG', 'EMG', 'EYEGAZE', 'GSR', 'HEOG', 'MISC', 'PUPIL', 'RESP', 'TEMP', 'VEOG', // physio
+    'AUDIO', 'REF', 'SYSCLOCK', 'TRIG', 'ADC', 'DAC', // system
+    'MEGMAG', 'MEGGRADAXIAL', 'MEGGRADPLANAR', 'MEGREFMAG', 'MEGREFGRADAXIAL', 'MEGOTHER', 'HLU', 'FITERR' // MEG
```
Comment (sappelhoff) [inline bids-validator/validators/tsv/checkTypeCol.js]: From what I see, this is the complete list:

>{'MEGGRADPLANAR', 'ECG', 'EOG', 'REF', 'GSR', 'EYEGAZE', 'MEGGRADAXIAL', 'PUPIL', 'ECOG', 'AUDIO', 'MEGREFMAG', 'MEGREFGRADPLANAR', 'TRIG', 'EMG', 'MEGREFGRADAXIAL', 'VEOG', 'SYSCLOCK', 'ADC', 'TEMP', 'RESP', 'EEG', 'HEOG', 'FITERR', 'OTHER', 'HLU', 'SEEG', 'MISC', 'PD', 'MEGOTHER', 'DBS', 'MEGMAG', 'DAC'}


So one is missing from your list


```suggestion
    'MEGMAG', 'MEGGRADAXIAL', 'MEGGRADPLANAR', 'MEGREFMAG', 'MEGREFGRADAXIAL', 'MEGREFGRADPLANAR', 'MEGOTHER', 'HLU', 'FITERR' // MEG
```

File: bids-validator/validators/tsv/checkTypeCol.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+]
+
+
+/**
+ * Checks type column in a (m/i) electroencephalography _channels.tsv file to
```
Comment (sappelhoff) [inline bids-validator/validators/tsv/checkTypeCol.js]:
```suggestion
 * Checks type column in an ephys _channels.tsv file to
```

File: bids-validator/validators/tsv/checkTypeCol.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+                        code: 130,
+                    }),
+                )
+            }
+        }
+    }
+    return
+}
+
+export default checkTypeCol
```
Comment (sappelhoff) [inline bids-validator/validators/tsv/checkTypeCol.js]: Let's try to end files with blank new lines, see: https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-a-source-file

image

```suggestion
export default checkTypeCol

```

**Review by sappelhoff — APPROVED** [review]
awesome! Thanks a lot @adam2392 

I just think that given the developments in https://github.com/bids-standard/bids-specification/issues/534, we should wait a bit with merging and potentially adjust the PR to always error when the thing is not uppercase.

cc @jasmainak @rwblair

File: bids-validator/utils/issues/list.js [hunk 7]
```diff
@@ -713,5 +713,15 @@ export default {
     key: 'SCANS_FILENAME_NOT_MATCH_DATASET',
     severity: 'error',
     reason: 'The filename in scans.tsv file does not match what is present in the BIDS dataset.'
+  },
+  130: {
+    key: 'CHANNELS_COLUMN_TYPE_UPPER_CASE',
+    severity: 'warning',
+    reason: 'Type column in channels.tsv files should consist of upper-case characters.'
```
Comment (jasmainak) [inline bids-validator/utils/issues/list.js:720]: is this tested somewhere? That the warning is raised?

Comment (sappelhoff) [reply bids-validator/utils/issues/list.js:720]: Not yet @jasmainak but we may even drop this warning and always error if it's not *exactly* as in the list (i.e., upper case)

Comment (sappelhoff) [reply bids-validator/utils/issues/list.js:720]: so I think let's wait for the discussion in https://github.com/bids-standard/bids-specification/issues/534 to finish, and then adjust this PR accordingly

File: bids-validator/tests/tsv.spec.js [hunk 8]
```diff
@@ -487,12 +485,57 @@ describe('TSV', function() {
...
+  it('should not allow channels.tsv files with lower-casing in type column', function() {
+    var tsv =
+      'name\ttype\tunits\tlow_cutoff\thigh_cutoff\tstatus\n' +
+      'value-name\teeg\tmV\tvalue-lowcut\tvalue-highcut\tgood'
+    validate.TSV.TSV(channelsFileEEG, tsv, [], function(issues) {
+      assert(issues.length === 1 && issues[0].code === 130)
+    })
+  })
+
+  it('should allow iEEG channels.tsv files with value other than accepted values in type column', function() {
```
Comment (sappelhoff) [inline bids-validator/tests/tsv.spec.js]:
```suggestion
  it('should allow iEEG channels.tsv files with accepted values in type column', function() {
```

**Comment by adam2392** [comment]
Not sure why the current CI is failing tbh...

**Comment by sappelhoff** [comment]
perhaps @rwblair can help?

**Review by sappelhoff — APPROVED** [review]
Ah I see now, test errors are related to the HED npm package, they'll be fixed elsewhere. Your tests covering the new code all pass and I think this is nice, so I am going to merge it. Thanks @adam2392

**Pull request merged**